File: java/pgjdbc/src/main/java/software/amazon/dsql/examples/alternatives/no_connection_pool/ExampleWithNoConnector.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

package software.amazon.dsql.examples.alternatives.no_connection_pool;

import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
import software.amazon.awssdk.services.dsql.DsqlUtilities;
import software.amazon.awssdk.services.dsql.model.GenerateAuthTokenRequest;
import software.amazon.awssdk.regions.Region;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.Properties;

public class ExampleWithNoConnector {

    // Get a connection to Aurora DSQL.
    public static Connection getConnection(String endpoint, String user, String region) throws SQLException {
        DsqlUtilities utilities = DsqlUtilities.builder()
                .region(Region.of(region))
                .credentialsProvider(DefaultCredentialsProvider.builder().build())
                .build();

        // The token expiration time is optional, and the default value is 900 seconds
        GenerateAuthTokenRequest tokenGenerator = GenerateAuthTokenRequest.builder()
                .hostname(endpoint)
                .region(Region.of(region))
                .build();

        // Generate a fresh password token for each connection, to ensure the token is
        // not expired when the connection is established
        String password;
        if (user.equals("admin")) {
            password = utilities.generateDbConnectAdminAuthToken(tokenGenerator);
        } else {
            password = utilities.generateDbConnectAuthToken(tokenGenerator);
        }

        Properties props = new Properties();
        props.setProperty("user", user);
        props.setProperty("password", password);
        // Use the DefaultJavaSSLFactory so that Java's default trust store can be used
        // to verify the server's root cert.
        props.setProperty("sslmode", "verify-full");
        props.setProperty("sslfactory", "org.postgresql.ssl.DefaultJavaSSLFactory");
        props.setProperty("sslNegotiation", "direct");

        String url = "jdbc:postgresql://" + endpoint + ":5432/postgres";

        return DriverManager.getConnection(url, props);
    }

    public static void main(String[] args) throws SQLException {
        String clusterEndpoint = System.getenv("CLUSTER_ENDPOINT");
        assert clusterEndpoint != null : "CLUSTER_ENDPOINT environment variable is not set";

        String clusterUser = System.getenv("CLUSTER_USER");
        assert clusterUser != null : "CLUSTER_USER environment variable is not set";

        String region = System.getenv("REGION");
        assert region != null : "REGION environment variable is not set";

        try (Connection conn = ExampleWithNoConnector.getConnection(clusterEndpoint, clusterUser, region)) {
            if (!clusterUser.equals("admin")) {
                try (Statement setSchema = conn.createStatement()) {
                    setSchema.execute("SET search_path=myschema");
                }
            }
            // Create a new table named owner
            try (Statement create = conn.createStatement()) {
                create.executeUpdate("""
                        CREATE TABLE IF NOT EXISTS owner(
                        id uuid NOT NULL DEFAULT gen_random_uuid(),
                        name varchar(30) NOT NULL,
                        city varchar(80) NOT NULL,
                        telephone varchar(20) DEFAULT NULL,
                        PRIMARY KEY (id))""");
            }

            // Insert some data
            try (PreparedStatement insert = conn.prepareStatement(
                    "INSERT INTO owner (name, city, telephone) VALUES (?, ?, ?)")) {
                insert.setString(1, "John Doe");
                insert.setString(2, "Anytown");
                insert.setString(3, "555-555-1999");
                insert.executeUpdate();
            }

            // Read back the data and assert they are present
            try (PreparedStatement read = conn.prepareStatement("SELECT * FROM owner WHERE name = ?")) {
                read.setString(1, "John Doe");
                try (ResultSet rs = read.executeQuery()) {
                    while (rs.next()) {
                        assert rs.getString("id") != null;
                        assert rs.getString("name").equals("John Doe");
                        assert rs.getString("city").equals("Anytown");
                        assert rs.getString("telephone").equals("555-555-1999");
                    }
                }
            }

            // Delete some data
            try (PreparedStatement delete = conn.prepareStatement("DELETE FROM owner WHERE name = ?")) {
                delete.setString(1, "John Doe");
                delete.executeUpdate();
            }
        }
        System.out.println("Connection exercised successfully");
    }
}

File: javascript/postgres-js/src/alternatives/no_connection_pool/example_with_no_connector.js
import { DsqlSigner } from "@aws-sdk/dsql-signer";
import postgres from "postgres"

import assert from "node:assert";

const ADMIN = "admin";
const PUBLIC = "public";
const NON_ADMIN_SCHEMA = "myschema";

async function getConnection(clusterEndpoint, user, region) {
  
  let client = postgres({
    host: clusterEndpoint,
    user: user,
    // We can pass a function to password instead of a value, which will be triggered whenever
    // connections are opened.
    password: async () => await getPasswordToken(clusterEndpoint, user, region),
    database: "postgres",
    port: 5432,
    idle_timeout: 2,
    ssl: {
      rejectUnauthorized: true,
    }
    // max: 1, // Optionally set maximum connection pool size
  })

  return client;
}

async function getPasswordToken(clusterEndpoint, user, region) {
  const signer = new DsqlSigner({
    hostname: clusterEndpoint,
    region,
  });
  if (user === ADMIN) {
    return await signer.getDbConnectAdminAuthToken();
  }
  else {
    signer.user = user;
    return await signer.getDbConnectAuthToken()
  }
}

async function example() {
  let client;

  const clusterEndpoint = process.env.CLUSTER_ENDPOINT;
  assert(clusterEndpoint);
  const user = process.env.CLUSTER_USER;
  assert(user);
  const region = process.env.REGION;
  assert(region);
  
  try {
    
    client = await getConnection(clusterEndpoint, user, region)
    let schema = user === ADMIN ? PUBLIC : NON_ADMIN_SCHEMA;

    // Note that due to connection pooling, we cannot execute 'set search_path=myschema'
    // because we cannot assume the same connection will be used.
    await client`CREATE TABLE IF NOT EXISTS ${client(schema)}.owner (
      id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
      name VARCHAR(30) NOT NULL,
      city VARCHAR(80) NOT NULL,
      telephone VARCHAR(20)
    )`;

    // Insert some data
    await client`INSERT INTO ${client(schema)}.owner(name, city, telephone) VALUES('John Doe', 'Anytown', '555-555-0150')`

    // Check that data is inserted by reading it back
    const result = await client`SELECT id, city FROM ${client(schema)}.owner where name='John Doe'`;
    assert.deepEqual(result[0].city, "Anytown")
    assert.notEqual(result[0].id, null)

    // Delete data we just inserted
    await client`DELETE FROM ${client(schema)}.owner where name='John Doe'`

  } catch (error) {
    console.error(error);
    throw error;
  } finally {  
    await client?.end();
  }
}

export { example }
